API: Fix cython exception handling for exported extern C functions

seberg opened the pull request:
This hopefully fixes them for all the functions currently in the `.pyd` files.  A surprising amount of them look like scary thing I wouldn't mind to just delete :).

Anyway, had looked at some Cython code today and remembered this.

Closes gh-19291

seberg committed:
API: Fix cython exception handling for exported extern C functions

This hopefully fixes them for all the functions currently in the `.pyd`
files.  A surprising amount of them look like scary thing I wouldn't
mind to just delete :).

Anyway, had looked at some Cython code today and remembered this.

Closes gh-19291

eric-wieser commented on numpy/__init__.cython-30.pxd:
```diff
@@ -559,11 +559,11 @@ cdef extern from "numpy/arrayobject.h":
+    int PyArray_CastTo (ndarray, ndarray) except -1
+    int PyArray_CastAnyTo (ndarray, ndarray) except -1
+    int PyArray_CanCastSafely (int, int)  # writes errors
+    npy_bool PyArray_CanCastTo (dtype, dtype)  # writes errors
```
I think all these "writes errors" should either be `except *` (if you can't say anything about the return value), or `except? 0` (if errors always produce zero, but not all zeros are errors).

seberg replied:
Oh, "writes errors" here means it uses `PyErr_WriteUnraiseable`.  So I think for cython that should be as good as nothing.

eric-wieser replied:
Oh, I thought this meant "sets the exception but doesn't indicate it in the return value".

charris replied:
PyErr_WriteUnraiseable -> PyErr_WriteUnraisable

Just to help searching for the documentation :)

eric-wieser commented on numpy/__init__.cython-30.pxd:
```diff
@@ -587,34 +587,34 @@ cdef extern from "numpy/arrayobject.h":
     npy_intp PyArray_PyIntAsIntp (object)
-    int PyArray_Broadcast (broadcast)
+    int PyArray_Broadcast (broadcast) except -1
     void PyArray_FillObjectArray (ndarray, object)
```
Should this one (and probably a handful of other void functions) be `except *`? I assume this doesn't swallow errors.

seberg replied:
Arg, I didn't scan the `void` ones carefully, I guess.  This isn't designed to error, but clearly can...  (one of those bad functions).  (the xincref/decref family is also weird, I think they tend to clear, but probably not always... and they shouldn't be used anway, especially in cython)

seberg committed:
API: Add "except *" in cython `.pyd` where necessary based on review

The incref/decref function shouldn't be able to fail (especially the
decref).  But right now they can, this will be fixed when we redo
clearing (see gh-22924)

seberg committed:
DOC: Add release note for cython exception fixes

seberg commented:
As confirmed by Stefan Behnel in the issue, I think this should be rather safe to do.  Cython will insert error handling and any manual error handling simply becomes obsolete.

charris commented:
Let's give this a shot. Thanks Sebastian.

zoj613 commented:
Thanks @seberg

seberg commented:
@scoder is it still relevant to push such changes to cython (in which case I can make a PR)?
